### PR TITLE
winch: Epoch-based interruption

### PR DESCRIPTION
**Pull request by saulecabrera**
Closes https://github.com/bytecodealliance/wasmtime/issues/8091

This commit introduces support for epoch interruption to Winch. The heuristics around epoch check emission are identical to Cranelift's except for the fact that the current implementation doesn't introduce a local-based cache for the current epoch deadline. This is an intentional decision given Winch's focus on compilation performance. However, if needed in the future, knobs could be introduced to optionally introduce a local cache at the cost of reduced compilation performance.

### CODE REVIEW

File: winch/codegen/src/visitor.rs [hunk 1]
```diff
@@ -1717,6 +1717,11 @@ where
         if self.tunables.consume_fuel {
             self.emit_fuel_check();
         }
+
+        // Emit epoch check right after binding the loop header.
+        if self.tunables.epoch_interruption {
+            self.emit_epoch_check();
+        }
```
Comment (fitzgen) [inline winch/codegen/src/visitor.rs:1724]: Maybe move this `if self.tunables.epoch_interruption` into the `emit_epoch_check` helper? (And rename it to `maybe_emit_epoch_check` or something like that? Feel free to bikeshed.)

This code motion would simplify callers, which are all doing the exact same conditional.

File: winch/codegen/src/codegen/mod.rs [hunk 2]
```diff
@@ -1022,6 +1026,93 @@ where
...
+        let new_epoch = self.env.builtins.new_epoch::<M::ABI, M::Ptr>();
+
+        // Checks for runtime limits (e.g., fuel, epoch) are special since they
+        // require inserting abritrary function calls and control flow.
+        // Special care must be taken to ensure that all invariants are met. In
+        // this case, since `new_epoch` takes an argument and returns a value,
+        // we must ensure that any registers used to hold the current epoch
+        // value and deadline are not going to be needed later on by the
+        // function call.
+        let (epoch_deadline_var, epoch_counter_var) = self.context.without::<(Reg, Reg), M, _>(
```
Comment (fitzgen) [inline winch/codegen/src/codegen/mod.rs:1042]: Is the `_var` suffix fairly established in Winch? If so, ignore the following.

If not, then I find it a little surprising. I thought maybe there was a `Var` type that abstracts over registers and stack slots or something. Or introduces a virtual register concept or something like that. I think a `_reg` suffix would have avoided me going down that line of thought and inquiry, which would make it slightly easier to follow what's going on, as a code reader.

File: winch/codegen/src/codegen/mod.rs [hunk 3]
```diff
@@ -1022,6 +1026,93 @@ where
...
+        self.masm.bind(cont);
+
+        self.context.free_reg(epoch_deadline_var);
+        self.context.free_reg(epoch_counter_var);
+    }
+
+    fn emit_load_epoch_deadline_and_counter(
+        &mut self,
+        epoch_deadline_var: Reg,
+        epoch_counter_var: Reg,
```
Comment (fitzgen) [inline winch/codegen/src/codegen/mod.rs:1081]: And then I'd expect these to also be `_reg` if the last suggestion is taken.

File: winch/codegen/src/codegen/mod.rs [hunk 4]
```diff
@@ -1022,6 +1026,93 @@ where
         );
     }
 
+    /// Emit series of instructions that check the current epoch against its
+    /// deadline.
+    pub fn emit_epoch_check(&mut self) {
+        let cont = self.masm.get_label();
```
Comment (fitzgen) [inline winch/codegen/src/codegen/mod.rs:1032]: I think a one-line comment describing what this label is could also be helpful to code readers.

**Review by fitzgen — APPROVED** [review]
Looks great! Just a couple nitpicks inline.

Also I know that most of the testing will be covered by fuzzing and the existing epoch tests, but it may make sense to throw in a single disas test with epochs enabled as well that contains a some `if`s, `block`s, and a `loop` just to show that we insert epoch checks only on the loop headers and body prologue. Could also be used to see that deadlines are actually cached, if we ever get around to doing that here.

**Comment by saulecabrera** [comment]
@fitzgen agreed with all the suggestions,thanks! I've also opened https://github.com/bytecodealliance/wasmtime/pull/9737, which applies the suggestions in this PR to fuel checks.

**Pull request merged**